### PR TITLE
client: support Root CA rotation on server side

### LINKED ISSUE #11555: ETCD cannot handle ca bundles for peer-trusted-ca-file or trusted-ca-file
Etcd cannot handle cert bundles in the `peer-trusted-ca-file` or `trusted-ca-file` section. Without the ability to handle CA bundles, it is impossible to do a 0 downtime approach to CA rotation without resigning all active client and server certs at once.

If a CA bundle was allowed: A new CA could be created and made valid in all components in the first interation. Then client certs can be resigned with the new CA since the server components have the new CA plus the old CA in it's trust bundle. Once all clients have been resigned and downloaded the old + new CA the server components can be signed with the new CA and then the old CA can be effectively removed.

It appears this was meant to be fixed but I am able to replicate the issue in an etcd deployment today.
I will expose all the certs and command line configurations in this issue so the exact steps can be replicated.

### PR DESCRIPTION
**Pull request by yishuT**
* In TlsConfig creation, use `GetConfigForClient` instead of `GetCertificate`, so that we can load the latest CA.
* Add a go routine to refresh tls config from given cert/key paths, such reload loop can be enabled through flags.
* Change TlsInfo to pointer receiver to avoid unnecessary copy
* Fix all TlsInfo pass by values because sync.Once cannot be copied. Provide TlsInfo.Clone to copy TlsInfo
* Unit tests and integration tests

Fixes #11555

### CODE REVIEW

**Comment by yishuT** [comment]
Not sure about if we still need e2e tests.  In the integration test, server start and client conn with new certs is already tested.

File: client/pkg/transport/listener_test.go [hunk 1]
```diff
@@ -575,3 +588,182 @@ func TestSocktOptsEmpty(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	tr := &http.Transport{TLSClientConfig: cfg}
+	cli := &http.Client{Transport: tr}
+	go func() {
+		cli.Get("https://" + ln.Addr().String())
+	}()
+
+	conn, err := ln.Accept()
```
Comment (serathius) [inline client/pkg/transport/listener_test.go:726]: Will this block infinitely if there is a bug? Could we avoid infinite blocking in tests, for example have some timeout?

File: client/pkg/transport/listener.go [hunk 2]
```diff
@@ -501,12 +513,88 @@ func (info TLSInfo) ServerConfig() (*tls.Config, error) {
 	// and in TLS 1.3, cipher suites are not configurable
 	// setting Max TLS version to TLS 1.2 for go 1.13
 	cfg.MaxVersion = tls.VersionTLS12
-
 	return cfg, nil
 }
 
+// baseConfig is called on initial TLS handshake start.
+//
+// Previously,
```
Comment (serathius) [inline client/pkg/transport/listener.go:521]: Looks like you are describing the fix, however comments should reflect current state of the code. Could you possibly add this information to commit message.

File: client/pkg/transport/listener.go [hunk 3]
```diff
@@ -501,12 +513,88 @@ func (info TLSInfo) ServerConfig() (*tls.Config, error) {
 	// and in TLS 1.3, cipher suites are not configurable
 	// setting Max TLS version to TLS 1.2 for go 1.13
 	cfg.MaxVersion = tls.VersionTLS12
-
 	return cfg, nil
 }
 
+// baseConfig is called on initial TLS handshake start.
```
Comment (serathius) [inline client/pkg/transport/listener.go:519]: This function is no longer called `baseConfig`

File: client/pkg/transport/listener.go [hunk 4]
```diff
@@ -402,82 +482,14 @@ func (info TLSInfo) baseConfig() (*tls.Config, error) {
...
-	cs := make([]string, 0)
-	if info.TrustedCAFile != "" {
-		cs = append(cs, info.TrustedCAFile)
-	}
-	return cs
-}
-
-// ServerConfig generates a tls.Config object for use by an HTTP server.
-func (info TLSInfo) ServerConfig() (*tls.Config, error) {
+func (info *TLSInfo) serverConfig() (*tls.Config, error) {
```
Comment (serathius) [inline client/pkg/transport/listener.go:488]: Why we have two separate ways to get serverConfig? `serverConfig` and `ServerConfig`. It's a common pattern to have public/private function to call one from another to expose different interface, however I don't see any way that those functions share any code here. I think we should try to simplify code by trying to reuse config creation more.

File: client/pkg/transport/listener.go [hunk 5]
```diff
@@ -564,6 +652,10 @@ func (info TLSInfo) ClientConfig() (*tls.Config, error) {
 	return cfg, nil
 }
 
+func (info *TLSInfo) Close() {
```
Comment (serathius) [inline client/pkg/transport/listener.go:655]: Aside of integration tests I didn't find any other places where this is called. I would expect EtcdServer to call this. Is this correct or maybe I missed it?

**Comment by serathius** [comment]
Can you please cleanup commits? There are over 22 commits (most of them called "fix ..."), could you squash them and maybe rename them so they could be used for review?

File: client/pkg/transport/listener.go (hunk 5 above)
Comment (yishuT) [reply client/pkg/transport/listener.go:650]: fixed

**Comment by yishuT** [comment]
@serathius ping on this. Happy NY

**Comment by yishuT** [comment]
ping again. 

I think we need this feature anyway as root ca rotation in one of the fundamental operations

**Comment by serathius** [comment]
Sorry for delayed response, It's in my review queue. Can you fix the conflicts and rebase the change?

**Comment by yishuT** [comment]
chang to Draft until I fix the broken tests

**Comment by yishuT** [comment]
Finally make tests passed. I think the one failed is unrelated to my change

**Comment by yishuT** [comment]
hey, @serathius just a friendly ping

File: tests/integration/root_ca_rotation_test.go [hunk 6]
```diff
@@ -0,0 +1,233 @@
+package integration
```
Comment (serathius) [inline tests/integration/root_ca_rotation_test.go:1]: This is major feature, we should have both integration and e2e tests. Please take a look into https://github.com/etcd-io/etcd/issues/13637.on how to write one code that with run both.

**Comment by yishuT** [comment]
Hey, that's totally fine. Thanks for the review.

**Comment by yishuT** [comment]
@serathius qq, to test root CA rotation, I need to create clients certs sign by different root CAs. The current `Cluster` interface in [test framework](https://sourcegraph.com/github.com/etcd-io/etcd/-/blob/tests/framework/interface.go?L32) does not allow me to do that. Therefore, I might need to make change to the interface.  Do you prefer it to be in this PR which already changes many pieces or in a separate PR?

**Comment by yishuT** [comment]
Not stale

**Comment by serathius** [comment]
cc @spzala @ahrtr to help review

**Comment by ahrtr** [comment]
Thanks @yishuT for the PR and sorry for the late response. 

I will take a look after you resolve the conflict.

**Comment by yishuT** [comment]
thanks. Will try to get to this PR by end of next week

**Comment by yishuT** [comment]
@ahrtr please review

**Comment by ahrtr** [comment]
Sorry for the late response, I am working on https://github.com/etcd-io/etcd/issues/14370 right now. This PR will be a priority when I finish that one. 

I do not get time to review this PR so far, but 1K+ lines of code change seems too big to me. Please consider to break the PR if possible. Will have a closer look later.

**Comment by yishuT** [comment]
> Sorry for the late response, I am working on #14370 right now. This PR will be a priority when I finish that one.
> 
> I do not get time to review this PR so far, but 1K+ lines of code change seems too big to me. Please consider to break the PR if possible. Will have a closer look later.

this PR touches the TlsInfo, so I need to make change everywhere...

**Comment by yishuT** [comment]
@ahrtr  I have a proposal to break this PR into 
1. TlsInfo -> *TlsInfo so that it's not copied everywhere
2. Introduce the root CA rotation logic

sg?

**Comment by yishuT** [comment]
@ahrtr  hey, I would like to finish this PR. As discussed before, this PR is too big to review. I am going to split this pr into couple parts.
1. Make TlsInfo -> *TlsInfo. The reason is that to reload root ca, we need to have a background goroutine. 
2. do the real root ca rotation.

wdyt?

**Comment by hongbin** [comment]
Hi,

I saw this PR has been opened for two years without significant progress. I wonder if anything I can help to move forward this PR.

From my side, I am from Azure Kubernetes Service team. We do have requirement to rotate etcd CA in zero-downtime. This PR seems to be able to solve the problem. If the community needs additional help to merge this PR, please don't hesitate to ping me.

**Comment by yishuT** [comment]
I can resume the work if the community can review the PR. I didn't receive any response earlier if they like my proposal.  maybe @ahrtr ?

**Comment by hongbin** [comment]
We can split the PR into two as @ahrtr mentioned. @yishuT you need help with that?

**Comment by yishuT** [comment]
> We can split the PR into two as @ahrtr mentioned. @yishuT you need help with that?

I don't have time to work on it recently. Feel free to take it over

**Comment by hongbin** [comment]
> Fixes #11555

Thanks. After an analysis on this PR, I think we can split this PR in this way:
* Provide implementation for "GetConfigForClient". This will allow server CA rotation to work in term of functionality.
* Optimize the loading of CA files. In particular, avoid loading the CA files repeatedly on each request.

I am working on the first part: https://github.com/etcd-io/etcd/pull/16500

**Comment by aauren** [comment]
@yishuT 

Why did this issue get closed? The problem still exists and all of the PRs that were created are still outstanding waiting for maintainer review.

**Comment by yishuT** [comment]
> @yishuT 
> 
> 
> 
> Why did this issue get closed? The problem still exists and all of the PRs that were created are still outstanding waiting for maintainer review.


This pr is pending review for yrs. also I don't have time to work on it recently while there is another pr trying to solve the same problem so I closed this pr

**Comment by aauren** [comment]
My bad, I thought this was the primary issue tracking the problem not a PR. I should have paid more attention.

Thanks for trying to make this better for all of us. I wish that it had been accepted back when you had opened it.

**Comment by hongbin** [comment]
It is unfortunate that this feature didn't make any progress for years. It is quite frustrating to see the slowness on the code reviewing process.

**Pull request closed without merging**